### PR TITLE
Endosymbiosis lite

### LINKED ISSUE #3936: Implement endosymbiosis lite
endosymbiosis lite is a system whereby engulfing prokaryotes the player can unlock organelles based on what the eaten bacteria had. On top of the base functionality there should be at least some tutorial somewhere letting the player know that this is possible. Maybe triggered when the player engulfs a cell for the first time?

Related to unlocking organelles: https://github.com/Revolutionary-Games/Thrive/issues/3935

### PR DESCRIPTION
**Pull request by hhyyrylainen**
**Brief Description of What This PR Does**

I'm finally doing the endosymbiosis feature: https://forum.revolutionarygamesstudio.com/t/upgrades-unlocks-endosymbiosis-master-thread/915/90

**Related Issues**

closes #3936

**Progress Checklist**

Note: before starting this checklist the PR should be marked as non-draft.

- [x] PR author has checked that this PR works as intended and doesn't
      break existing features:
      https://wiki.revolutionarygamesstudio.com/wiki/Testing_Checklist
      (this is important as to not waste the time of Thrive team
      members reviewing this PR)
- [x] Initial code review passed (this and further items should not be checked by the PR author)
- [ ] Functionality is confirmed working by another person (see above checklist link)
- [ ] Final code review is passed and code conforms to the 
      [styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md).

Before merging all CI jobs should finish on this PR without errors, if
there are automatically detected style issues they should be fixed by
the PR author. Merging must follow our
[styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md#git).

### CODE REVIEW

**Comment by CI09** [comment]
Even though I used cheats to speed up testing, it seemed to be a really fun feature to play with.
When endosymbiosis is done, the candidate species wasn't present in the patch anymore (maybe intentional).
Also it takes 5 editor enters to fully complete the endosymbiosis process (at least for me); it can be consulted with some theorist how long did endosymbiosis take (if it is known) on Earth and just do some number tweak.
Very cool feature either way 👏

File: src/general/EndosymbiosisData.cs [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+                {
+                    if (endosymbiont.OriginallyFromSpecies.ID == targetSpecies)
+                    {
+                        GD.PrintErr("Trying to form symbiosis with something that is already an endosymbiont");
+                        return false;
+                    }
+                }
+            }
+
+            StartedEndosymbiosis = new InProgressEndosymbiosis(candidate.Key, cost, organelle);
```
Comment (CI09) [inline src/general/EndosymbiosisData.cs:100]: I'm probably wrong, but this is never discarded, so...

**Comment by hhyyrylainen** [comment]
> the candidate species wasn't present in the patch anymore (maybe intentional).

This would be a pretty tricky GUI part, so I decided that this will be opened as an issue to be done later (or at least highlight species not present in current patch).

> Also it takes 5 editor enters to fully complete the endosymbiosis process (at least for me);

The time it takes depends on how many organelles that can be turned into the endosymbiote there are in the species, so a species with a bunch of metabolosomes can be turned into a mitochondrion in just 2 cycles (so the potential range is 2-5 editor cycles).

File: src/general/EndosymbiosisData.cs (hunk 1 above)
Comment (hhyyrylainen) [reply src/general/EndosymbiosisData.cs:100]: What do you mean by discarded? This value is stored in the property and this method next returns true so the value will stay there until cancelled or completed.

**Comment by CI09** [comment]
This would be a pretty tricky GUI part, so I decided that this will be opened as an issue to be done later (or at least highlight species not present in current patch).

I mean, after I finished endosymbiosis, endosymbiont wasn't in patch anymore (like if it has gone extinct).

**Comment by hhyyrylainen** [comment]
> I mean, after I finished endosymbiosis, endosymbiont wasn't in patch anymore (like if it has gone extinct).

That's probably just a random chance it happened. There's no logic to delete a species that becomes an endosymbiote. It doesn't affect the target species at all that someone made an endosymbiote from it.

File: src/general/EndosymbiosisData.cs (hunk 1 above)
Comment (CI09) [reply src/general/EndosymbiosisData.cs:100]: > What do you mean by discarded?
Disposed. You know, it was created with `new` but isn't destroyed anywhere (I saw that `StartedEndosymbiosis` is being set to null somewhere though). But as I said, I am probably wrong here...

Comment (hhyyrylainen) [reply src/general/EndosymbiosisData.cs:100]: `InProgressEndosymbiosis` is not an `IDisposable` so it is not possible to even dispose it. As it is a plain C# class that just uses memory and references other similar objects, the garbage collector takes care of deallocating it when it is no longer used (which is kind of the point with garbage collected programming languages like C# versus C++).

**Pull request merged**